### PR TITLE
feat: Make SingleConsumerMultiProducer the default mailbox for stream.

### PR DESCRIPTION
**Pull request by He-Pin**
Motivation:
Make stream actor using the SingleConsumerMultiProducer  by default for performance.

refs: https://github.com/apache/incubator-pekko/issues/915

### CODE REVIEW

File: stream/src/main/resources/reference.conf [hunk 1]
```diff
@@ -21,6 +21,14 @@ pekko {
       # or full dispatcher configuration to be used by ActorMaterializer when creating Actors.
       dispatcher = "pekko.actor.default-dispatcher"
 
+      # FQCN of the MailboxType. The Class of the FQCN must have a public
+      # constructor with
+      # (org.apache.pekko.actor.ActorSystem.Settings, com.typesafe.config.Config) parameters.
+      # defaults to the single consumber mailbox for better performance.
```
Comment (pjfanning) [inline stream/src/main/resources/reference.conf:27]: misspelling - should be 'consumer'

File: stream/src/main/scala/org/apache/pekko/stream/impl/ActorMaterializerImpl.scala [hunk 2]
```diff
@@ -62,7 +62,8 @@ import pekko.util.OptionVal
     val effectiveProps = props.dispatcher match {
       case Dispatchers.DefaultDispatcherId =>
         // the caller said to use the default dispatcher, but that can been trumped by the dispatcher attribute
-        props.withDispatcher(context.effectiveAttributes.mandatoryAttribute[ActorAttributes.Dispatcher].dispatcher)
+        val dispatcher = context.effectiveAttributes.mandatoryAttribute[ActorAttributes.Dispatcher].dispatcher
+        props.withDispatcher(dispatcher)
```
Comment (pjfanning) [inline stream/src/main/scala/org/apache/pekko/stream/impl/ActorMaterializerImpl.scala:66]: can this be reverted?

File: stream/src/main/scala/org/apache/pekko/stream/impl/PhasedFusingActorMaterializer.scala [hunk 3]
```diff
@@ -62,6 +62,8 @@ import pekko.util.OptionVal
 
   val Debug = false
 
+  val Mailbox: String = "pekko.stream.materializer.mailbox"
```
Comment (pjfanning) [inline stream/src/main/scala/org/apache/pekko/stream/impl/PhasedFusingActorMaterializer.scala:65]: can you call this param 'MailboxConfigName' ?

File: stream/src/main/scala/org/apache/pekko/stream/impl/ActorMaterializerImpl.scala (hunk 2 above)
Comment (He-Pin) [reply stream/src/main/scala/org/apache/pekko/stream/impl/ActorMaterializerImpl.scala:66]: Why?

File: stream/src/main/scala/org/apache/pekko/stream/impl/PhasedFusingActorMaterializer.scala (hunk 3 above)
Comment (He-Pin) [reply stream/src/main/scala/org/apache/pekko/stream/impl/PhasedFusingActorMaterializer.scala:65]: Can be done.

File: stream/src/main/scala/org/apache/pekko/stream/impl/ActorMaterializerImpl.scala (hunk 2 above)
Comment (He-Pin) [reply stream/src/main/scala/org/apache/pekko/stream/impl/ActorMaterializerImpl.scala:66]: Seems I missed the mailbox part

Comment (pjfanning) [reply stream/src/main/scala/org/apache/pekko/stream/impl/ActorMaterializerImpl.scala:66]: the new code is the same as the old code - just 2 lines and a new 'val' - can't it just be left as it was?

Comment (laglangyue) [reply stream/src/main/scala/org/apache/pekko/stream/impl/ActorMaterializerImpl.scala:66]: line 65 and line 66 can be merged，no need to define the `val dispathcer`

File: stream/src/main/scala/org/apache/pekko/stream/impl/ActorMaterializerImpl.scala [hunk 4]
```diff
@@ -195,10 +197,14 @@ private[pekko] class SubFusingActorMaterializerImpl(
...
  */
 @InternalApi private[pekko] object StreamSupervisor {
-  def props(attributes: Attributes, haveShutDown: AtomicBoolean): Props =
+  def props(attributes: Attributes, haveShutDown: AtomicBoolean): Props = {
+    val dispatcher = attributes.mandatoryAttribute[ActorAttributes.Dispatcher].dispatcher
     Props(new StreamSupervisor(haveShutDown))
       .withDeploy(Deploy.local)
-      .withDispatcher(attributes.mandatoryAttribute[ActorAttributes.Dispatcher].dispatcher)
+      .withDispatcher(dispatcher)
+      .withMailbox(PhasedFusingActorMaterializer.MailboxConfigName)
```
Comment (laglangyue) [inline stream/src/main/scala/org/apache/pekko/stream/impl/ActorMaterializerImpl.scala:205]: Here, it seems that the dispatcher can rewrite through attributes.
Is mailbox necessary? I am not sure.

File: stream/src/main/scala/org/apache/pekko/stream/impl/ActorMaterializerImpl.scala (hunk 2 above)
Comment (He-Pin) [reply stream/src/main/scala/org/apache/pekko/stream/impl/ActorMaterializerImpl.scala:66]: Inline is too long, But if you both prefer, I can revert to old style.

I can do that tomorrow.

File: stream/src/main/scala/org/apache/pekko/stream/impl/PhasedFusingActorMaterializer.scala [hunk 5]
```diff
@@ -62,6 +62,8 @@ import pekko.util.OptionVal
 
   val Debug = false
 
+  val MailboxConfigName: String = "pekko.stream.materializer.mailbox"
```
Comment (laglangyue) [inline stream/src/main/scala/org/apache/pekko/stream/impl/PhasedFusingActorMaterializer.scala:65]: This is hard coding, I don't know if there is a place for  processing configaration specially?
like  `org.apache.pekko.stream.ActorAttributes#IODispatcher` ?

**Comment by laglangyue** [comment]
It seems that a default configuration has been added. Are there any unit test?

File: stream/src/main/scala/org/apache/pekko/stream/impl/PhasedFusingActorMaterializer.scala (hunk 5 above)
Comment (He-Pin) [reply stream/src/main/scala/org/apache/pekko/stream/impl/PhasedFusingActorMaterializer.scala:65]: It's read by typesafe config

File: stream/src/main/scala/org/apache/pekko/stream/impl/ActorMaterializerImpl.scala (hunk 4 above)
Comment (He-Pin) [reply stream/src/main/scala/org/apache/pekko/stream/impl/ActorMaterializerImpl.scala:205]: This code is right dude.

**Comment by He-Pin** [comment]
@laglangyue I don't this this code need any test, and a test may need another pr first:)

**Comment by laglangyue** [comment]
lgtm

**Comment by He-Pin** [comment]
@mdedetrich Would you like to give this a review? this was landed in Akka 2.7

**Review by mdedetrich — APPROVED** [review]
lgtm. It may be worth it to create a simple unit test that confirms `PhasedFusingActorMaterializer.MailboxConfigName` is actually being used when a stream is run, but if this is too hard then don't worry about it.

**Comment by He-Pin** [comment]
@mdedetrich Not easy to test, that's why I want the https://github.com/apache/incubator-pekko/pull/919 . and I do confirmed that works with debuger.

**Comment by mdedetrich** [comment]
> @mdedetrich Not easy to test, that's why I want the #919 . and I did confirm that works with debuger.

No worries, I approved the PR anyways so its good from myend

**Pull request merged**